File: LEC17/LEC17_02.py
import sys
largest_node = input()
graph = {}
while True:
    input_ = sys.stdin.readline()
    if input_ == '':
        break
    u, v = input_[0], input_[1]
    if u not in graph:
        graph[u] = set()
    graph[u].add(v)
    if v not in graph:
        graph[v] = set()
    graph[v].add(u)


def dfs(graph, start, visited):
    visited.add(start)
    for node in graph[start]:
        if node not in visited:
            dfs(graph, node, visited)
    return visited


visited = dfs(graph, largest_node, set())
print(len(visited))

File: LEC17/LEC17_02.cpp
#include <bits/stdc++.h>

using namespace std;

int main()
{
  char largest_node;
  cin >> largest_node;
  map<char, set<char>> graph;
  while (true)
  {
    char u, v;
    cin >> u >> v;
    if (cin.eof())
      break;
    graph[u].insert(v);
    graph[v].insert(u);
  }

  set<char> visited;
  queue<char> q;
  // count the number of components
  int count = 0;
  for (char u = 'A'; u <= largest_node; u++)
    if (visited.find(u) == visited.end())
    {
      count++;
      q.push(u);
      while (!q.empty())
      {
        char u = q.front();
        q.pop();
        visited.insert(u);
        for (char v : graph[u])
          if (visited.find(v) == visited.end())
            q.push(v);
      }
    }
  cout << count << endl;
}